Have VectorImpl use static allocation for small sizes

knelli2 opened the pull request:
## Proposed changes

Note to reviewers: The test is mostly whitespace changes, so hiding whitespace changes may be useful.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Have VectorImpl use static allocation for small sizes

wthrowe commented on src/DataStructures/VectorImpl.hpp:
```diff
@@ -119,12 +128,12 @@ struct MarkAsVectorImpl {};
-template <typename T, typename VectorType>
+template <typename T, typename VectorType,
+          size_t StaticSize = default_vector_impl_static_size>
 class VectorImpl
```
The TensorExpression code, at least, assumes `VectorImpl` only has two parameters.  For example, this won't match custom-sized vectors:
```c++
template <typename ValueType, typename VectorType>
struct get_binop_datatype_impl<VectorImpl<ValueType, VectorType>, ValueType> {
```

wthrowe commented on src/DataStructures/VectorImpl.hpp:
```diff
@@ -297,43 +297,56 @@ class VectorImpl
-          rhs.size() > 0
-              ? cpp20::make_unique_for_overwrite<value_type[]>(rhs.size())
-              : nullptr) {
+/// \cond HIDDEN_SYMBOLS
```
Why hide these?

knelli2 replied:
I was getting errors when building the docs that Doxygen wasn't able to find some symbols. The declarations are already hidden above, so I just hid the definitions as well.

wthrowe reviewed (changes requested):
I would like more justification for this change.  What problem is it trying to solve?

macedo22 committed:
Support custom VectorImpl static sizes in TensorExpressions

knelli2 commented:
Added a new commit that addresses the concerns about TensorExpressons (big thanks to @macedo22). 

As for justification, the main reason that prompted me to add this was the functions of time. Every time we call `func`, we do a heap allocation for the `DataVector`(s) and we compute the time dependent part of the maps pretty often so we were doing a bunch of heap allocations. However, for almost all functions of time (except for shape) we know the number of components, the number won't change throughout the evolution, and each one has 4 components or less. With these changes, now we do stack allocations for the `DataVector`s in the functions of time.

Since `DataVector` is technically a `NodalVector`, it shouldn't really be used for the functions of time. So in a follow-up PR I was going to add a new vector type that we use for the functions of time.

knelli2 commented:
Test failure seems unrelated

wthrowe commented on tests/Unit/DataStructures/Tensor/Expressions/CustomStaticSizeVector.hpp:
```diff
@@ -0,0 +1,33 @@
+// Distributed under the MIT License.
```
Move this file into `tests/Unit/Helpers/`.

wthrowe reviewed (changes requested):
OK, that's a reasonable use case.

A few more files that I think need updating:
- docs/DevGuide/ImplementingVectors.md (probably just the last section)
- src/DataStructures/DynamicBuffer.hpp
- src/DataStructures/Variables.hpp

knelli2 commented:
@wthrowe Ok I think I got everything. Let me know if I missed anything

nilsdeppe commented on src/DataStructures/VectorImpl.hpp:
```diff
@@ -83,6 +83,10 @@ constexpr bool is_assignable_v = is_assignable<LhsDataType, RhsDataType>::value;
 
+/// \ingroup DataStructuresGroup
+/// Default static size for vector impl
+constexpr size_t default_vector_impl_static_size = 5;
```
Should we keep this at `0` for now? I'm not sure it will benefit us except maybe in 1d, and if we know we need it, we can use the `HybridVector` you said you'll add. It'll also help keep down memory usage. @wthrowe @kidder thoughts on this?

wthrowe replied:
I like this idea.

wthrowe commented on docs/DevGuide/ImplementingVectors.md:
```diff
@@ -324,25 +328,37 @@ memory. Non-owning vectors do not manage memory, nor can they change size. The
-`blaze::CustomVector` internal pointer to the pointer obtained by
-`std::unique_pointer.get()`.
+`blaze::CustomVector` internal pointer to either the pointer obtained by
+`std::unique_pointer.get()` or the pointer obtained by `std::array.get()`
```
`array.data()`

wthrowe commented on docs/DevGuide/ImplementingVectors.md:
```diff
@@ -324,25 +328,37 @@ memory. Non-owning vectors do not manage memory, nor can they change size. The
-`set_data_ref` function, the internal `std::unique_ptr` named `owned_data_` no
-longer points to the data represented by the vector and can be thought of as
+`set_data_ref` function, neither the internal `std::unique_ptr` named
+`owned_data_` nor the internal `std::array` names `static_owned_data_`
```
"named"

wthrowe commented on src/DataStructures/DynamicBuffer.hpp:
```diff
@@ -28,14 +28,19 @@ class er;
 class DynamicBuffer {
  public:
+  // Rather than having to template DynamicBuffer on the static size, just use
+  // the default static size since this only determines if something is derived
+  // from VectorImpl and not the specific size itself
   static constexpr bool is_data_vector_type =
-      std::is_base_of_v<VectorImpl<tt::get_fundamental_type_t<T>, T>, T>;
+      std::is_base_of_v<VectorImpl<tt::get_fundamental_type_t<T>, T,
+                                   default_vector_impl_static_size>,
+                        T>;
```
I don't understand.  Isn't this checking that the static size is the default?

knelli2 replied:
Yes. So currently, `DynamicBuffer` is only used in `YlmSpherepack`. And `DynamicBuffer` is only instantiated for `double, DataVector`. No other types. So we want to check if it's using the default because that's what `DataVector` uses. If in the future we want to extend `DynamicBuffer` to other vector types that don't use the default static size, then we can do that.

wthrowe replied:
If it's only supposed to work for DataVector just check if it's a DataVector instead of doing stuff with VectorImpl inheritance.

knelli2 commented:
Pushed some fixups setting the default static size to 0 and addressing other comments

wthrowe commented on tests/Unit/Helpers/DataStructures/VectorImplTestHelper.hpp:
```diff
@@ -44,15 +44,20 @@ void vector_test_construct_and_assign(
-  std::vector<size_t> sizes{static_sdist(gen), default_vector_impl_static_size,
-                            sdist(gen)};
+  std::vector<size_t> sizes;
+  if constexpr (default_vector_impl_static_size >= 2) {
```
Instead of checking the default size, check the size for `VectorType` and call the test functions somewhere with a vector with non-zero static size.

nilsdeppe reviewed (commented):
I have no additional comments

knelli2 commented:
Edited the existing fixups. Added a new test with a custom vector with non-zero static size

wthrowe commented:
OK, I'm happy now.

knelli2 commented:
Squashed and rebased.

nilsdeppe commented:
CI failure is fixed in #4289
